### DISCUSSION TITLE
How To Listen For Things Using Gram-JS

### DISCUSSION THREAD

**Discussion by Roebou1989** [gram-js/gramjs#467] [General]
Hi Everyone,

I have ran through so many pages and sites over the past couple of days, and would really like to request some help. I have done a Discord bot before, where I do the following, which I would really like to be able to do something similar here for Telegram, but I seem to keep on being stuck.

Discord Bot 1:
1. Setup bot credentials
2. Use bot.connect()
3. Use bot.on('ready', run async function on an interval to fetch participants)
4. Automatically fetch participants in channels

Discord Bot 2:

1. Setup bot credentials
2. Use bot.connect()
3. Use bot.once('ready', async function to setup commands)
4. Use bot.on('interactionCreate', async function to handle new messages being sent)

As can be seen, I connect my bot in discord and then I can run something once off, like when the bot is ready from the ready event and then I can also run specific actions when there is a message ('interactionCreate') event specific to my bot.

How do I go about something like this for my Telegram bot?

High level steps:

1. Setup credentials (have this working)
2. Connect the bot 
3. Run a once off function when the bot has been started / connected
4. Run a function every time a new message is sent, for example, /register 12345 , and then do something in the function and reply 'Thanks, registered'.

Thanks in advance for anyone willing to assist!
I am using Javascript / NodeJS as my backend.

Thanks,
Rohan

**Comment by rojvv** [comment]
This is a demonstration of a GramJS exquivalent of what you’re talking about:

```ts
import { NewMessage } from "telegram/events";

client.addEventHandler(async (event) => {
	// Use the args from event.message.patternMatch.
	await event.message.reply({ text: "Thanks, registered!" });
}, new NewMessage({ pattern: /^\/register(@yourBotUsername|) ([0-9]+)$/ }));

await client.start({ ... });
// Call your async function to setup commands.
```

Note that GramJS is communicating directly to MTProto API which provides a lower-level interface for building Telegram clients. If you focus mainly on bots, and want to build highly extensible ones, you might want to use a framework like [grammY](https://github.com/grammyjs/grammy).

**Reply by Roebou1989** [reply]
Thanks for the response @roj1512 .

I have tried the below code snippet, but I get an error when running it.

````
require('dotenv').config();
require('console-stamp')(console, {
  format: ':label :date(yyyy/mm/dd HH:MM:ss.l,true)',
});

const { Api, TelegramClient } = require("telegram");
const { StringSession } = require("telegram/sessions");
const { NewMessage } = require("telegram/events");
const { db } = require('./models');

const session = new StringSession(process.env.SESSION_ID);
const client = new TelegramClient(session, parseInt(process.env.API_ID), process.env.API_HASH, {});

async function main() {

  client.addEventListener(async (event) => {
    // Use the args from event.message.patternMatch.
    await event.message.reply({ text: "Thanks, registered!" });
  }, new NewMessage({ pattern: /^\/add(@yourBotUsername|) ([0-9]+)$/ }));
  
  await client.connect();
  
}

main();
````

````
[LOG] [2023/01/16 19:13:13.173] [2023-01-16T21:13:13.171] [INFO] - [Running gramJS version 2.15.3]
/Users/roebou/Documents/GitHub/tillies-telegram-bot-nodejs/bot2.js:16
  client.addEventListener(async (event) => {
         ^

TypeError: client.addEventListener is not a function
    at main (/Users/roebou/Documents/GitHub/tillies-telegram-bot-nodejs/bot2.js:16:10)
    at Object.<anonymous> (/Users/roebou/Documents/GitHub/tillies-telegram-bot-nodejs/bot2.js:25:1)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:17:47
[nodemon] app crashed - waiting for file changes before starting...
````

**Reply by rojvv** [reply]
Seems like I mistyped the method name. It should be `client.addEventHandler`.

**Reply by rojvv** [reply]
See the edited snippet.

**Reply by Roebou1989** [reply]
I tried the addEventHandler as well, but nothing ended up happening.
I have tried out your suggestion of Grammy, which works much better for me and a lot like the Discord bot, which is what I am comfortable with.

The main issue that I am sitting is that I want to get the members of a group or channel, as I need to store this information on my side. 

1. Get all groups and channels that have added my bot as an administrator
2. Get all users for each of the groups and channels and save them against my user profiles.

Do you know how I can perhaps do this with the alternative library? Apologies for all the questions, going through this bit by bit.

Thanks so much for your assistance, much appreciated!

**Reply by rojvv** [reply]
> The main issue that I am sitting is that I want to get the members of a group or channel, as I need to store this information on my side.

You can’t directly get all of the groups that your bot is member of, regardless of the library you’re using. The only solution for this is to use a database, updating records whenever your bot receives a message.

> Thanks so much for your assistance, much appreciated!

You’re always welcome!
